jdb/assignment2/tree.py

```diff
@@ -0,0 +1,196 @@
+   they are available when needed, and they are only computed once:
+   - Nodes keep a pointer to the parent
+   - Nodes keep a int representing the depth of a Node in the tree.
+
+   If the tree was mutable, we could still make sure mutations would keep these
+   2 signals correct, but it would be more work.
+
+2. the exercise requests no storing of the nodes: this code does not create any
+   list to which nodes get appended, instead, this code uses generators which
+   are comparable to index pointer being moved. They are baked into the Python
+   language with the yield keyword so they are less verbose than in other
+   languages like C++ or Java.
+
+By construction, the two tree properties are guaranteed: connected graph and no
+loops. In addition, one large assumption in this code, is that values held by
+the nodes are unique. The unit test uses ints, there are no 2 nodes with the
+value 2, for example.
+
+Additional note about Python type annotations: Python 3.5 introduced optional
+typing, and types complement unit tests:
+1. they help the readers of the code,
+2. they help the author to design a good consistent API,
+3. type checkers find errors statically, thus saving some class of errors
+   before the code is sent to production.
+4. static analysis helps with refactoring, and refactoring is a fact of life for
+    medium to big sized project.
+We suggest and warmly support the use of Python type annotations :)
+More info at: https://docs.python.org/3/library/typing.html
+
+To run the mypy static type checker (http://mypy-lang.org/):
+$ mypy tree.py
+"""
+
+import unittest
+from typing import Generator, Tuple, TypeVar
+
+T = TypeVar('T')  # the generic type of values held in each node.
+
+class Node:
+  def __init__(self, value : T,
+               left: Tuple = None,
+               right: Tuple = None,
+               parent: 'Node' = None,
+               depth: int = 0):
+
+    self.value = value
+    self.parent = parent
+    self.depth = depth
+    self.left = Node(*left, parent=self, depth=depth+1) if left else None
+    self.right = Node(*right, parent=self, depth=depth+1) if right else None
+
+  def _ancestors(self) -> Generator['Node', None, None]:
+    """Yields the current node, then the ancestor nodes up to the root.
+
+    The first element yield is the node itself (in this sense, the node is
+    considered an ancestor of itself).
+    """
+    yield self
+    if self.parent:
+      yield from self.parent._ancestors()
+
+  def print_ancestors(self):
+    print(' '.join(str(i) for i in self._ancestors()))
+
+  def lowest_common_ancestor(self, node1: 'Node', node2: 'Node') -> 'Node':
+    """Returns the node that is the lowest common ancestor of node1, node2."""
+    deeper, shallower  = ((node2, node1)
+                          if node2.depth > node1.depth else (node1, node2))
+    for _ in range(deeper.depth - shallower.depth):
+      deeper = deeper.parent
```

Comment:
This is a smart way to speed up the function when nodes are at very different depths, I will remember this approach!